rfcs/20190726-custom-ops.md

```diff
@@ -0,0 +1,56 @@
+# Best practices for custom operations in TensorFlow
+
+| Status        | Proposed                                          |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Alexandre Passos (apassos@google.com |
+| **Sponsor** | Karmel Allison (karmel@google.com) |
+| **Updated**   | 2019-06-10                                           |
+
+For most of TF’s history, it was very expensive for third-party packages or
+libraries to release their own tf operations. This created pressure to put ops
+in tf core or in tf contrib, which created some uncertainty around support
+stories and backwards compatibility.
+
+Around (but technically not a part of) TF 2.0, however, TensorFlow supports [a
+straightforward way for third-party package to build and deploy their own custom
+TF ops](https://github.com/tensorflow/custom-op/blob/master/README.md). To
+maintain a healthy ecosystem, we recommend the following best practices.
+
+## Experimental ops should live out of tree
+
+Unless some special considerations apply, experimental op development should not
+happen inside the core TensorFlow package. Strongly prefer adding experimental
+or new operations to libraries and packages downstream from core TensorFlow. Any
+op in core TensorFlow is subject to very strict backward and forward
+compatibility policies, as TensorFlow is very aggressive about not breaking
+existing GraphDefs, and this includes even meant-to-be experimental operations
+in the core TensorFlow package.
+
+Once things are no longer experimental, and once the TensorFlow team determines
+it is ok with taking responsibility for the code, it’s fine to propose adding a
+new version with the final intended interface and implementation to core
+TensorFlow. The intermediate states are best explored in another package.
+
+This has many advantages:
+ - downstream packages often have a faster release cadence than core TensorFlow
+ - each downstream package can choose its own backward and forward compatibility
+   processes, allowing fine-grained trade-offs between velocity and stability
+
+## Out-of-tree ops must be namespaced
+
+Since an op’s name uniquely identifies it, different TF packages should ensure
+their op names are globally unique across the entire TF ecosystem. To do so,
+prepend the package’s name to the op’s name and separate with a ‘>’. An op named
+“MatMul” inside the “tensorflow_addons” package should be named “Addons>MatMul”,
+for example.
+
+The string used for a package’s component name is any valid op name, but should
```

Comment:
It would be convenient to developers if you can include a link to the documentation of what constitutes a valid op name (a regex?)

Reply:
The regex is in https://github.com/tensorflow/tensorflow/blob/876f07dbc27ce790b076291c74550538ad183c45/tensorflow/core/framework/op_def.proto#L17

Reply:
In that case, isn't the suggestion above "To do so, prepend the package’s name to the op’s name and separate with a ‘>’."  not going to work, given that the regex appears to be "[A-Z][a-zA-Z0-9_]*"?

Reply:
Clearly this proposal should be read as implying a change to this regex to be something like ([A-Z][a-zA-Z0-9_]>)$EXISTING_REGEX

Reply:
OK. Thanks for the clarification.